### PR TITLE
feat(memory): add typed compact stage seam

### PR DESCRIPTION
**Pull request by KasumiChen**
## Summary

- add a typed built-in compact-stage seam around pre-compaction durable flush so compaction-owned memory work reports `StageDiagnostics` under `MemoryStageFamily::Compact`
- route `maybe_compact_context(...)` through the central memory compact-stage entrypoint while preserving current fail-open behavior and runtime-owned compaction policy
- add regression coverage for compact-stage diagnostics so the built-in contract distinguishes succeeded and skipped compact-stage outcomes without widening the current runtime boundary

## Scope

- [x] Small and focused
- [ ] Includes docs updates (if needed)
- [x] No unrelated refactors

Scope note: this is a follow-up slice after `#464`, limited to the compaction-owned durable-flush path and its typed stage contract.

## Risk Track

- [ ] Track A (routine/low-risk)
- [x] Track B (higher-risk/policy-impacting)

Risk note: this changes the runtime-to-memory contract on the pre-compaction path, so the main risk is preserving the existing fail-open and skip behavior exactly.

## Validation

- [x] `/home/kasumi/.cargo/bin/cargo fmt --all -- --check`
- [x] `/home/kasumi/.cargo/bin/cargo clippy --workspace --all-targets --all-features -- -D warnings`
- [x] `/home/kasumi/.cargo/bin/cargo test --workspace --all-features --locked`
- [x] Additional scenario/benchmark checks (if applicable)
- [ ] If this changes config/env fallback, limits, or defaults: include before/after behavior and regression coverage for explicit path, fallback path, and boundary values
- [ ] If tests mutate process-global env: document how state is restored or serialized

Additional checks executed:

- [x] `/home/kasumi/.cargo/bin/cargo test -p loongclaw-app compact_stage -- --nocapture`
- [x] `/home/kasumi/.cargo/bin/cargo test -p loongclaw-app handle_turn_with_runtime_flushes_durable_memory_before_compaction -- --nocapture`
- [x] `/home/kasumi/.cargo/bin/cargo test -p loongclaw-app handle_turn_with_runtime_does_not_flush_durable_memory_when_compaction_is_skipped -- --nocapture`
- [x] `/home/kasumi/.cargo/bin/cargo test -p loongclaw-app maybe_compact_context_fails_open_when_durable_flush_cannot_write_workspace_export -- --nocapture`
- [x] `/home/kasumi/.cargo/bin/cargo test -p loongclaw-app default_runtime_kernel_build_context -- --nocapture`
- [x] `/home/kasumi/.cargo/bin/cargo test --workspace --locked`

## Linked Issues

Refs #455

## Summary by CodeRabbit

* **Refactor**
  * Improved memory compaction flow with richer diagnostics and controlled fail‑open behavior for clearer outcomes and more reliable recovery.
* **Tests**
  * Added tests covering compaction diagnostics: success, skipped conditions, and repeated-run behavior.
* **Documentation**
  * Updated architecture drift metadata and tracked metrics for the memory module.

### CODE REVIEW

File: crates/app/src/memory/orchestrator.rs [hunk 1]
```diff
@@ -253,6 +253,51 @@ fn run_rank_stage(entries: Vec<MemoryContextEntry>) -> StageRunResult {
...
+
+async fn run_builtin_compact_stage(
+    session_id: &str,
+    workspace_root: Option<&Path>,
+    config: &MemoryRuntimeConfig,
+) -> Result<StageDiagnostics, String> {
+    match super::flush_pre_compaction_durable_memory(session_id, workspace_root, config).await {
+        Ok(super::durable_flush::PreCompactionDurableFlushOutcome::Flushed { .. })
+        | Ok(super::durable_flush::PreCompactionDurableFlushOutcome::SkippedDuplicate) => {
+            Ok(StageDiagnostics::succeeded(MemoryStageFamily::Compact))
```
[the remark this answers is not in the document]
Comment (KasumiChen) [reply crates/app/src/memory/orchestrator.rs]: Fixed in `79322c1`.

Changes:
- map `PreCompactionDurableFlushOutcome::SkippedDuplicate` to `StageOutcome::Skipped`
- add `compact_stage_skips_when_durable_flush_is_duplicate`
- keep the existing succeeded path only for a real `Flushed` outcome

Verified with:
- `$HOME/.cargo/bin/cargo test -p loongclaw-app compact_stage -- --nocapture`
- `$HOME/.cargo/bin/cargo test -p loongclaw-app maybe_compact_context_fails_open_when_durable_flush_cannot_write_workspace_export -- --nocapture`

**Comment by KasumiChen** [comment]
Follow-up patch is up at `79322c1`.

Additional changes beyond the inline duplicate-skip fix:
- crate-scope `memory::run_compact_stage` export so the coordinator-only seam does not widen the public surface
- add a non-`memory-sqlite` compact-stage fallback path so this slice no longer hard-references `durable_flush` symbols when that feature is off

Verification for this patch:
- `$HOME/.cargo/bin/cargo fmt --all -- --check`
- `$HOME/.cargo/bin/cargo test -p loongclaw-app compact_stage -- --nocapture`
- `$HOME/.cargo/bin/cargo test -p loongclaw-app maybe_compact_context_fails_open_when_durable_flush_cannot_write_workspace_export -- --nocapture`
- `$HOME/.cargo/bin/cargo test -p loongclaw-app handle_turn_with_runtime_flushes_durable_memory_before_compaction -- --nocapture`
- `$HOME/.cargo/bin/cargo test -p loongclaw-app handle_turn_with_runtime_does_not_flush_durable_memory_when_compaction_is_skipped -- --nocapture`
- `cargo check -p loongclaw-app --no-default-features` still fails elsewhere in this tree, but its output no longer includes `memory/orchestrator.rs`, `durable_flush`, or `flush_pre_compaction_durable_memory` references

**Review by chumyin — APPROVED** [review]
hi @KasumiChen, thanks for the follow-up here. i refreshed the march architecture drift snapshot so governance matches the current  footprint, reran the compact-stage regression coverage, reran , and rechecked the updated ci path. this is in good shape now. nice work tightening the compact-stage seam, and congrats on getting this slice ready to land.

**Review by chumyin — APPROVED** [review]
hi @KasumiChen, thanks for the follow-up here. i refreshed the march architecture drift snapshot so governance matches the current memory_mod footprint, reran the compact-stage regression coverage, reran cargo test --workspace --all-features --locked, and rechecked the updated ci path. this is in good shape now. nice work tightening the compact-stage seam, and congrats on getting this slice ready to land.
